### PR TITLE
IC-1324 / IC-1326 / IC-1328 cancel a referral with a reason

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

- add contract and service layer method for cancelling a referral

- add contract and service layer method for getting referral cancellation reason

## What is the intent behind these changes?

allow referrals to be cancelled by the PP with a specific reason

### CODE REVIEW

File: server/services/interventionsService.ts [hunk 1]
```diff
@@ -52,6 +52,15 @@ export interface SentReferral {
   endOfServiceReport: EndOfServiceReport | null
 }
 
+export interface EndedReferral {
+  id: string
+  referenceNumber: string
+  referral: ReferralFields
+  endedAt: string
+  endedBy: AuthUser
+  cancellationReason: string
```
Comment (Gweaton) [inline server/services/interventionsService.ts:61]: Are these arbitrary, or are there a list of them somewhere? e.g. is it worth specifying a stronger type/enum?

File: server/services/interventionsService.ts [hunk 2]
```diff
@@ -220,6 +220,11 @@ export interface UpdateDraftEndOfServiceReportParams {
   outcome: CreateEndOfServiceReportOutcome | null
 }
 
+export interface CancellationReason {
+  code: string
```
Comment (Gweaton) [inline server/services/interventionsService.ts:224]: see other comment re: use of `string` here

**Review by Gweaton — APPROVED** [review]
Looks good, just had one question

**Review by Gweaton — CHANGES_REQUESTED** [review]
You've asked me to remind you to add the "additional comments" here

File: server/services/interventionsService.ts (hunk 1 above)
Comment (tomsmyers) [reply server/services/interventionsService.ts:61]: there is a list but it's subject to change and not part of the contract as such i don't think.

they are three letter codes. i could define a type that's three letters explicitly?

File: server/services/interventionsService.test.ts [hunk 3]
```diff
@@ -2183,6 +2183,82 @@ pactWith({ consumer: 'Interventions UI', provider: 'Interventions Service' }, pr
...
+      await provider.addInteraction({
+        state: 'There is an existing sent referral with ID of 400be4c6-1aa4-4f52-ae86-cbd5d23309bf',
+        uponReceiving:
+          'a POST request to end the referral with ID 400be4c6-1aa4-4f52-ae86-cbd5d23309bf because SU was recalled ',
+        withRequest: {
+          method: 'POST',
+          path: '/sent-referral/400be4c6-1aa4-4f52-ae86-cbd5d23309bf/end',
+          body: {
+            cancellationReasonCode: 'REC',
+            additionalComments: 'Alex was arrested for driving without insurance and immediately recalled.',
```
Comment (tomsmyers) [inline server/services/interventionsService.test.ts:2211]: here

File: server/services/interventionsService.test.ts [hunk 4]
```diff
@@ -2183,6 +2183,82 @@ pactWith({ consumer: 'Interventions UI', provider: 'Interventions Service' }, pr
...
+      })
+    })
+
+    it('returns an ended referral', async () => {
+      expect(
+        await interventionsService.cancelReferral(
+          token,
+          '400be4c6-1aa4-4f52-ae86-cbd5d23309bf',
+          'REC',
+          'Alex was arrested for driving without insurance and immediately recalled.'
```
Comment (tomsmyers) [inline server/services/interventionsService.test.ts:2229]: here

File: server/services/interventionsService.ts [hunk 5]
```diff
@@ -52,6 +52,16 @@ export interface SentReferral {
...
 }
 
+export interface EndedReferral {
+  id: string
+  referenceNumber: string
+  referral: ReferralFields
+  endedAt: string
+  endedBy: AuthUser
+  cancellationReason: string
+  cancellationComments: string | null
```
Comment (tomsmyers) [inline server/services/interventionsService.ts:62]: here

File: server/services/interventionsService.ts [hunk 6]
```diff
@@ -604,4 +619,29 @@ export default class InterventionsService {
       headers: { Accept: 'application/json' },
     })) as EndOfServiceReport
   }
+
+  async cancelReferral(
+    token: string,
+    referralId: string,
+    reasonCode: string,
+    additionalComments: string
```
Comment (tomsmyers) [inline server/services/interventionsService.ts:627]: and here

File: server/services/interventionsService.ts [hunk 7]
```diff
@@ -604,4 +619,29 @@ export default class InterventionsService {
...
+  ): Promise<EndedReferral> {
+    const restClient = this.createRestClient(token)
+    return (await restClient.post({
+      path: `/sent-referral/${referralId}/end`,
+      data: {
+        cancellationReasonCode: reasonCode,
+        additionalComments,
```
Comment (tomsmyers) [inline server/services/interventionsService.ts:634]: oh, and here

**Comment by tomsmyers** [comment]
@rgforsyth added you here for visibility about the 'additionalComments' arg

**Review by Gweaton — APPROVED** [review]

**Pull request merged**